### ISSUE TITLE
Possible optimization for `detach` call in `_fnDraw`

### ISSUE THREAD

**Issue by xPaw** [DataTables/DataTablesSrc#264]
This code:
https://github.com/DataTables/DataTablesSrc/blob/5f976cf412efc60c3c3ef521f95c80b51b7ddbf6/js/core/core.draw.js#L444

is not super efficient because it causes jquery to call `removeChild` on every row in the body.

![image](https://github.com/DataTables/DataTablesSrc/assets/613331/aeb44c1d-cc34-41a6-b1ed-d9adb4b8184c)

Here's a hacky example that shows it could be faster:
```js
	var newBody = $('<tbody>');
	newBody.insertBefore(body);

	body.detach();

	body = newBody;
	oSettings.nTBody = body;
```

![image](https://github.com/DataTables/DataTablesSrc/assets/613331/1b3ca3b8-c8b5-4661-ae97-6b0e9012e478)

As you can see it goes from 1 second to 9ms on page load on a page with 9k rows. There's certainly an opportunity to optimize this.

**Comment by AllanJard** [comment]
Its a neat idea, but I don't think it is viable unfortunately, due to it being a new element. It would mean event handlers such as:

```js
$('#myTable tbody').on('click', 'tr', ...);
```

Would no longer work.

I'm astonished it takes almost a second to remove table rows - go you have a particularly large table which is being DOM loaded? I wonder if doing a native `removeChild` might be faster than a jQuery `detach`?

**Comment by xPaw** [comment]
I preload tables without using ajax yeah.

I understand making a new element isn't good here. Perhaps there's a fast way to clear out the  children. Maybe .innerHTML='' or replaceChildren call.

**Comment by AllanJard** [comment]
`replaceChildren` is an interesting option. Are you able to provide details, or a link to your page, so I can profile it and see what difference some changes might make.

**Comment by xPaw** [comment]
https://steamdb.info/sales/ is my page, but any table with a lot of rows in html should do the trick.


I initially hide the table and only show it after datatable initializes as that massively helps load times.

**Comment by xPaw** [comment]
`oSettings.nTBody.replaceChildren( ...anRows );` does work.

**Comment by xPaw** [comment]
I'm getting around 20ms for that call. There's still some other things I'd like to take a look at, as the whole loadedInit still takes ~500ms (down from 1.39ms without the detach change).

As a side note, while searching I've noticed that `_pagingDraw` is getting called twice because `DataTable.feature.register( 'paging', ...)` is getting called twice on page load when using the legacy `dom` option. Wouldn't consider this a big issue (I will look into migrating to `layout`), just something to think about.

**Comment by AllanJard** [comment]
If we switch to `replaceChildren`, it will need a little feature detection to make sure that is available. My goal with DataTables is to support browsers which are 10 years old and younger. That method only really became wide spread in 2020. It should also use `.apply()` rather than a spread operator, again for compatibility.

I'm happy to make that change, or take an updated PR as you wish.

Regarding the paging component - really good point. I've [committed a fix](https://github.com/DataTables/DataTablesSrc/commit/2fdf7acad6a634bd05dc6f5eef072a488ba640ac) to address that error.

**Comment by AllanJard** [comment]
I'm afraid not. I profiled it a fair bit while doing the DT2 dev work and I think I got most of the low hanging fruit. I might have missed something of course, but I'm not sure what. If you are able to post a test case for profiling that would be useful.

**Comment by xPaw** [comment]
The link I posted above is a good test case. From my profiling, it seems like there's a few things I've noticed

- Parsing data- attributes (like data-sort) if they're present, cellProcess is pretty expensive, partially because it does things like _fnSplitObjNotation

I have no good suggestions on how to improve this though.

- _fnFilterData is checking for presence of `&` in the contents, which puts it into an invisible div to then recover textContent, my question is why aren't cell contents retrieved as text to begin with in `cellProcess`?

**Comment by AllanJard** [comment]
Sorry, I missed the link! I'll take a look at it in the next few days.

I honestly can't recall off the top of my head why I elected to do it that way. I'd need to read over the code and comments. It might just be that I didn't think of another way, or it might have been required for the orthogonal data code paths.

**Comment by xPaw** [comment]
Here's a diff for the initially reported issue, with your suggestions:

```js
		if ( 'replaceChildren' in Element.prototype )
		{
			Element.prototype.replaceChildren.apply( body.get(0), anRows );
		}
		else
		{
			body.children().detach();
			body.append( $(anRows) );
		}
```

**Comment by AllanJard** [comment]
Thank you. I've [committed a change](https://github.com/DataTables/DataTablesSrc/commit/d1126e05f821e271b9754cdc57a4cfd711fd6327) to add the use of `replaceChildren`. All unit tests are passing with it :).

I'll close this issue for now, but if you think of any other performance improvements, just flag open a new one :)

**Comment by AllanJard** [comment]
Good point. I'll look into that, probably for 2.1.

**Comment by AllanJard** [comment]
I've been looking at this and the `innerHTML / textContent` would actually have an impact on not just DOM loaded content, but also on Ajax / JS loaded data, since it might have entities escaped and that would need to be handled. The orthogonal loading of the data might also introduce some complexities. I think if I were starting over, then I would go down that road, but at the moment, changing it would likely have a very significant impact.

Regarding the `cellProcess` other than perhaps a little bit of function caching I don't have much of an idea at the moment how to improve its performance. DOM interaction is always the slow part.
